### PR TITLE
[JUJU-860] SystemState to return error instead of nil if pool is closed

### PR DESCRIPTION
**Pull request by cderici**
#### Description

Currently the `SystemState` function in `state/pool.go` returns a `nil` whenever the pool is closed. This PR provides a semantics preserving (as mechanical as possible) change to have the `SystemState` function to return an error instead of a nil when the pool is closed.

Related [LP bug](https://bugs.launchpad.net/juju/+bug/1967330) reports that a client is getting a nil pointer de-refence seemingly coming from the `SystemState`. This change doesn't necessarily fix the LP bug by itself, but it's a step towards it. Regardless of that bug, we should have that function to return an error instead of nil if the pool is closed.

#### QA Steps

The CI tests should pass and the following should work fine

```sh
  $ juju bootstrap localhost test
  $ juju upgrade-controller --build-agent
```

as well as with `juju enable-ha`, e.g. :

```sh
  $ juju bootstrap localhost test
  $ juju enable-ha
  $ juju upgrade-controller --build-agent
```

should all be working ok.

### CODE REVIEW

File: apiserver/apiserver.go [hunk 1]
```diff
@@ -757,7 +766,8 @@ func (srv *Server) endpoints() []apihttp.Endpoint {
 			return opener, st, nil
 		},
 	}
-	controllerAdminAuthorizer := controllerAdminAuthorizer{srv.shared.statePool.SystemState()}
+	systemState, _ = srv.shared.statePool.SystemState()
```
Comment (juanmanuel-tirado) [inline apiserver/apiserver.go:769]: Process error just in case.

File: apiserver/facades/client/client/backend.go [hunk 2]
```diff
@@ -171,7 +171,8 @@ type poolShim struct {
 }
 
 func (p *poolShim) SystemState() *state.State {
-	return p.pool.SystemState()
+	systemState, _ := p.pool.SystemState()
```
Comment (juanmanuel-tirado) [inline apiserver/facades/client/client/backend.go:174]: Process error.

File: apiserver/watcher.go [hunk 3]
```diff
@@ -1051,7 +1051,8 @@ var getMigrationBackend = func(st *state.State) migrationBackend {
 }
 
 var getControllerBackend = func(pool *state.StatePool) controllerBackend {
-	return pool.SystemState()
+	systemState, _ := pool.SystemState()
```
Comment (juanmanuel-tirado) [inline apiserver/watcher.go:1054]: Process error.

File: state/controller.go [hunk 4]
```diff
@@ -54,7 +54,8 @@ func (ctlr *Controller) StatePool() *StatePool {
 
 // SystemState returns the State object for the controller model.
 func (ctlr *Controller) SystemState() *State {
-	return ctlr.pool.SystemState()
+	systemState, _ := ctlr.pool.SystemState()
```
Comment (juanmanuel-tirado) [inline state/controller.go:57]: Process the error.

File: upgrades/backend.go [hunk 5]
```diff
@@ -125,11 +125,13 @@ type stateBackend struct {
 }
 
 func (s stateBackend) ControllerUUID() string {
-	return s.pool.SystemState().ControllerUUID()
+	systemState, _ := s.pool.SystemState()
```
Comment (juanmanuel-tirado) [inline upgrades/backend.go:128]: Process the error.

File: upgrades/backend.go [hunk 6]
```diff
@@ -125,11 +125,13 @@ type stateBackend struct {
...
+	return systemState.ControllerUUID()
 }
 
 func (s stateBackend) StateServingInfo() (controller.StateServingInfo, error) {
-	return s.pool.SystemState().StateServingInfo()
+	systemState, _ := s.pool.SystemState()
```
Comment (juanmanuel-tirado) [inline upgrades/backend.go:133]: Process error.

File: upgrades/backend.go [hunk 7]
```diff
@@ -253,11 +259,16 @@ func (s stateBackend) RemoveContainerImageStreamFromNonModelSettings() error {
...
+	systemState, err := s.pool.SystemState()
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	return systemState.ControllerConfig()
 }
 
 func (s stateBackend) LeaseNotifyTarget(logger raftleasestore.Logger) raftlease.NotifyTarget {
-	return s.pool.SystemState().LeaseNotifyTarget(logger)
+	systemState, _ := s.pool.SystemState()
```
Comment (juanmanuel-tirado) [inline upgrades/backend.go:270]: Process the error.

**Comment by hmlanigan** [comment]
@cderici As this issue is more refrequenztly seen in HA, please add to the qa steps

File: apiserver/apiserver.go [hunk 8]
```diff
@@ -625,7 +633,8 @@ func (srv *Server) endpoints() []apihttp.Endpoint {
 		noModelUUID     bool
 	}
 	var endpoints []apihttp.Endpoint
-	controllerModelUUID := srv.shared.statePool.SystemState().ModelUUID()
+	systemState, _ := srv.shared.statePool.SystemState()
```
Comment (hmlanigan) [inline apiserver/apiserver.go:636]: another error to process, though I'm not sure why the same method is calling srv.shared.statePool.SystemState() again at ln 769

**Review by hmlanigan — CHANGES_REQUESTED** [review]
Failed QA
Run `(cd tests ; ./main.sh -v controller)` results in failure and unusable controller. 
```console
$ juju deploy ubuntu
ERROR not master and slaveOk=false
```

Code review ongoing.

File: upgrades/backend.go [hunk 9]
```diff
@@ -16,6 +15,7 @@ import (
 	"github.com/juju/juju/environs/config"
 	"github.com/juju/juju/state"
 	raftleasestore "github.com/juju/juju/state/raftlease"
+	"github.com/juju/replicaset/v2"
```
Comment (hmlanigan) [inline upgrades/backend.go:18]: this import is in the wrong stanza, belongs with juju/errors.

**Review by hmlanigan — COMMENTED** [review]
Initial review the code is done.

More testing is needed, especially for upgrades and ha.

Please add unit tests where method/function signatures change to return an error based on this change.

**Comment by hmlanigan** [comment]
@cderici thank you for this work! it's a big effort to change something that's used everywhere.

File: apiserver/facades/agent/caasoperator/register.go [hunk 10]
```diff
@@ -36,8 +36,12 @@ func newStateFacade(ctx facade.Context) (*Facade, error) {
 	if err != nil {
 		return nil, errors.Annotate(err, "getting leadership client")
 	}
+	systemState, err := ctx.StatePool().SystemState()
```
[the remark this answers is not in the document]
Comment (cderici) [reply apiserver/facades/agent/caasoperator/register.go:39]: No, it's a field of the StatePool and I think that guaranteeing that would be beyond the function of `SystemState()`

File: upgrades/backend.go (hunk 9 above)
Comment (cderici) [reply upgrades/backend.go:18]: @hmlanigan what did you mean by this? :point_up:

**Comment by juanmanuel-tirado** [comment]
!!build!!

**Comment by hmlanigan** [comment]
@cderici even with the new changes, I still hit the `ERROR not master and slaveOk=false` response. Please try `(cd tests ; ./main.sh -v controller)`

File: upgrades/backend.go (hunk 9 above)
Comment (hmlanigan) [reply upgrades/backend.go:18]: imports for juju are in 3 stanzas/groups, std go libs, non juju/juju libs and juju/juju libs. "github.com/juju/replicaset/v2" belongs in the 2nd.

**Comment by cderici** [comment]
> @cderici even with the new changes, I still hit the `ERROR not master and slaveOk=false` response. Please try `(cd tests ; ./main.sh -v controller)`

@hmlanigan thanks for QAing this, I don't seem to be getting any errors with `(cd tests ; ./main.sh -v controller)`. I double checked (re-run the tests after re-compiling with these changes), and this is the output that I got both the times: https://pastebin.canonical.com/p/Y26rT8qKwC/

lxd version : `4:24`

File: upgrades/backend.go (hunk 9 above)
Comment (cderici) [reply upgrades/backend.go:18]: Ah ok I didn't pay attention to the `juju/juju/` vs `juju/`, I'll fix it :+1: good catch!

File: apiserver/facades/agent/provisioner/provisioner.go [hunk 11]
```diff
@@ -34,6 +31,8 @@ import (
 	"github.com/juju/juju/state/watcher"
 	"github.com/juju/juju/storage"
 	"github.com/juju/juju/storage/poolmanager"
+	"github.com/juju/loggo"
+	"github.com/juju/names/v4"
```
Comment (hmlanigan) [inline apiserver/facades/agent/provisioner/provisioner.go:35]: these moved to the wrong section: https://github.com/juju/juju/blob/2.9/CONTRIBUTING.md#imports

File: apiserver/watcher.go [hunk 12]
```diff
@@ -1081,11 +1082,15 @@ func newMigrationStatusWatcher(context facade.Context) (facade.Facade, error) {
 	if !ok {
 		return nil, apiservererrors.ErrUnknownWatcher
 	}
+	cB, err := getControllerBackend(pool)
```
Comment (hmlanigan) [inline apiserver/watcher.go:1085]: A more descriptive variable name, than cB, would be helpful here

File: upgrades/backend.go [hunk 13]
```diff
@@ -254,11 +268,16 @@ func (s stateBackend) RemoveContainerImageStreamFromNonModelSettings() error {
...
+		return nil, errors.Trace(err)
+	}
+	return systemState.ControllerConfig()
 }
 
-func (s stateBackend) LeaseNotifyTarget(logger raftleasestore.Logger) raftlease.NotifyTarget {
-	return s.pool.SystemState().LeaseNotifyTarget(logger)
+func (s stateBackend) LeaseNotifyTarget(logger raftleasestore.Logger) (raftlease.NotifyTarget, error) {
+	systemState, err := s.pool.SystemState()
+	return systemState.LeaseNotifyTarget(logger), err
```
Comment (hmlanigan) [inline upgrades/backend.go:280]: Check the error before using systemState

File: upgrades/raft_test.go [hunk 14]
```diff
@@ -551,9 +551,9 @@ func (s *mockState) DropLeasesCollection() error {
 	return s.stub.NextErr()
 }
 
-func (s *mockState) LeaseNotifyTarget(logger raftleasestore.Logger) raftlease.NotifyTarget {
+func (s *mockState) LeaseNotifyTarget(logger raftleasestore.Logger) (raftlease.NotifyTarget, error) {
 	s.stub.AddCall("LeaseNotifyTarget", logger)
-	return s.target
+	return s.target, nil
```
Comment (hmlanigan) [inline upgrades/raft_test.go:556]: s/nil/s.stub.NextErr()/

File: upgrades/upgrade_test.go [hunk 15]
```diff
@@ -207,9 +207,9 @@ type mockStateBackend struct {
 	testing.Stub
 }
 
-func (mock *mockStateBackend) ControllerUUID() string {
+func (mock *mockStateBackend) ControllerUUID() (string, error) {
 	mock.MethodCall(mock, "ControllerUUID")
-	return "a-b-c-d"
+	return "a-b-c-d", nil
```
Comment (hmlanigan) [inline upgrades/upgrade_test.go:212]: s/nil/mock.stub.NextErr()/

**Comment by cderici** [comment]
!!build!!

**Comment by cderici** [comment]
!!build!!

**Comment by cderici** [comment]
!!build!!

**Comment by hmlanigan** [comment]
I forgot to update - the `not master and slaveOk=false` error is related to mongo replica set failovers when 2 of 3 nodes of HA are removed. This is a separate issue with a bug: https://bugs.launchpad.net/juju/+bug/1971627

**Review by hmlanigan — APPROVED** [review]
LGTM - great change

**Pull request merged**